src/chainstate/stacks/db/transactions.rs

```diff
@@ -0,0 +1,1001 @@
+                    checked_assets.insert(asset_id);
+                },
+                TransactionPostCondition::Nonfungible(ref asset_info, ref asset_value_str, ref condition_code) => {
+                    let asset_id = AssetIdentifier {
+                        contract_identifier: QualifiedContractIdentifier::new(StandardPrincipalData::from(asset_info.contract_address.clone()), asset_info.contract_name.clone()),
+                        asset_name: asset_info.asset_name.clone()
+                    };
+
+                    let asset_sent_condition = match asset_value_str.try_as_clarity_literal() {
+                        Some(value) => value,
+                        None => {
+                            return false;
+                        }
+                    };
+
+                    let empty_assets = vec![];
+                    let assets_sent = asset_map.get_nonfungible_tokens(&account.principal, &asset_id).unwrap_or(&empty_assets);
+                    if !condition_code.check(&asset_sent_condition, assets_sent) {
+                        debug!("Post-condition check failure on non-fungible asset {:?} owned by {:?}: {:?} {:?}", &asset_id, account, &asset_sent_condition, condition_code);
+                        return false;
+                    }
+
+                    checked_assets.insert(asset_id);
+                }
+            }
+        }
+
+        if !allow_unchecked_assets {
+            // make sure every asset transferred is covered by a postcondition
+            let mut fungible_asset_ids = asset_map.get_fungible_token_ids(&account.principal);
+            let mut nonfungible_asset_ids = asset_map.get_nonfungible_token_ids(&account.principal);
+            let stx_transfer_opt = asset_map.get_stx(&account.principal);
+
+            for asset_id in fungible_asset_ids.drain(..) {
+                if !checked_assets.contains(&asset_id) {
+                    debug!("Post-condition check failure on fungible asset {:?} owned by {:?}: missing post-condition check", &asset_id, account);
+                    return false;
+                }
+            }
+
+            for asset_id in nonfungible_asset_ids.drain(..) {
+                if !checked_assets.contains(&asset_id) {
+                    debug!("Post-condition check failure on non-fungible asset {:?} owned by {:?}: missing post-condition check", &asset_id, account);
+                    return false;
+                }
+            }
+
+            if stx_transfer_opt.is_some() && !checked_stx {
+                debug!("Post-condition check failure on STX: missing post-condition check");
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /// Process a token transfer payload (but wrapped in its transaction still, in order to do
+    /// post-condition checks).
+    fn process_transaction_token_transfer<'a>(clarity_tx: &mut ClarityTx<'a>, tx: &StacksTransaction, origin_account: &StacksAccount) -> Result<(), Error> {
```

Comment:
Since this function panic when it's supplied anything but a TokenTransfer, it'd be simpler to type the argument `tx` as `&TransactionTokenTransfer`, that way the type checker will enforce the match, rather than a check and panic.

Reply:
Ah, I didn't know you could do that.  Will change.

Related, I think we're dropping support for non-STX token transfers?  I might just remove most of this code outright.

Reply:
Okay, so I can't actually change this -- we need to pass the whole tx into this method in order to evaluate post-conditions.

Reply:
Ah, yes, I see. Somewhat relatedly, why would a transfer need post-conditions?

Reply:
I guess we don't need them for STX transfers -- it's not like they add any value in this particular case.  I'm fine with just requiring that these transactions have zero post-conditions.

Speaking of things that should or should not be allowed in STX transfers, I did think of a reason why you'd want to support sponsored STX transfers (which are currently prohibited) -- I would want to do this for transferring STX from code storage.  I would sign my STX transaction with my cold storage key on an air-gapped computer, and then use a separate (much-lower-value) account to sponsor the transaction and relay it safely.  Importantly, I don't need to guess the fee when I'm on the air-gapped computer -- I can wait for an arbitrarily long amount of time before sending the cold-storage-signed transaction, since the use of a separate sponsoring key effectively gives me a way to late-bind on the fee.  I went ahead and removed this prohibition in my tree.

Reply:
Agreed -- I think _any_ transaction should be sponsorable (similar to a sighash anyone can pay). Allowing a separate key to pay transaction fees is useful for any number of reasons.

Reply:
Addressed in d399b6d9113660aa1a71e88d9e7915c6dd8bb12f -- the transaction sighash for origin accounts no longer commits to anything in sponsor accounts (only that the transaction will be sponsored by someone), so a sponsor can sign a transaction already signed by the origin account and commit to its own nonce and tx fee (neither of which the origin knows).  This should be equivalent to sighash-anyone-can-pay.